lib/srv/handlers.go

```diff
@@ -0,0 +1,292 @@
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/sshutils"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// AuthHandlers are common authorization and authentication related handlers
+// used by the regular and forwarding server.
+type AuthHandlers struct {
+	// Server is the services.Server in the backend.
+	Server services.Server
+
+	// ProxyMode is if the server is running in proxy mode. Only applies to the
+	// regular server.
+	ProxyMode bool
+
+	// AuditLog is the service used to access Audit Log.
+	AuditLog events.IAuditLog
+
+	// AccessPoint is used to access the Auth Server.
+	AccessPoint auth.AccessPoint
+}
+
+// CheckAgentForward checks the certificate was signed by a valid certificate
+// authority, fetches the appropriate roles, and checks if agent forwarding
+// is allowed.
+func (h *AuthHandlers) CheckAgentForward(ctx *ServerContext) error {
+	cert, err := ctx.GetCertificate()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ca, err := h.authorityForCert(cert.SignatureKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	roles, err := h.fetchRoleSet(cert, ca, ctx.TeleportUser, ctx.ClusterName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := roles.CheckAgentForward(ctx.Login); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// KeyAuth implements SSH client authentication using public keys and is called
+// by the server every time the client connects
+func (h *AuthHandlers) KeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+	cid := fmt.Sprintf("conn(%v->%v, user=%v)", conn.RemoteAddr(), conn.LocalAddr(), conn.User())
+	fingerprint := fmt.Sprintf("%v %v", key.Type(), sshutils.Fingerprint(key))
+	log.Debugf("[SSH] %v auth attempt with key %v", cid, fingerprint)
+
+	logger := log.WithFields(log.Fields{
+		"local":       conn.LocalAddr(),
+		"remote":      conn.RemoteAddr(),
+		"user":        conn.User(),
+		"fingerprint": fingerprint,
+	})
+
+	certChecker := ssh.CertChecker{IsAuthority: h.isAuthority}
+
+	cert, ok := key.(*ssh.Certificate)
+	log.Debugf("[SSH] %v auth attempt with key %v, %#v", cid, fingerprint, cert)
+	if !ok {
+		log.Debugf("[SSH] auth attempt, unsupported key type for %v", fingerprint)
+		return nil, trace.BadParameter("unsupported key type: %v", fingerprint)
+	}
+	if len(cert.ValidPrincipals) == 0 {
+		log.Debugf("[SSH] need a valid principal for key %v", fingerprint)
+		return nil, trace.BadParameter("need a valid principal for key %v", fingerprint)
+	}
+
+	if len(cert.KeyId) == 0 {
+		log.Debugf("[SSH] need a valid key ID for key %v", fingerprint)
+		return nil, trace.BadParameter("need a valid key for key %v", fingerprint)
+	}
+	teleportUser := cert.KeyId
+
+	logAuditEvent := func(err error) {
```

Comment:
I would rename to `recordFailedLogin` and remove extra `if` as all the calls seem to be only for failed events